src/common/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp

```diff
@@ -38,10 +41,29 @@ bool ngraph::pass::SharedShapeOf::run_on_model(const std::shared_ptr<ngraph::Fun
     for (const auto& pair : source_to_shape_of) {
         if (pair.second.size() < 2)
             continue;
-        const auto& root_ss = pair.second[0];
-        for (const auto& child_ss : pair.second)
-            if (root_ss->get_instance_id() != child_ss->get_instance_id() && root_ss->get_output_element_type(0) == root_ss->get_output_element_type(0))
-                graph_rewritten |= replace_output_update_name(child_ss->output(0), root_ss->output(0));
+
+        std::vector<std::shared_ptr<ngraph::Node>> nodes_for_different_types[2];
+        for (const auto& child : pair.second) {
+            const auto child_as_shapeof_v8 =  std::dynamic_pointer_cast<opset8::ShapeOf>(child);
+            if (child_as_shapeof_v8) {
+                const auto& type_of_output = child_as_shapeof_v8->get_output_type();
+                size_t index = (type_of_output == element::i32) ? index_for_int32 : index_for_int64;
+                nodes_for_different_types[index].push_back(child);
+            } else {
+                const auto child_as_shapeof_v1 =  std::dynamic_pointer_cast<opset1::ShapeOf>(child);
+                if (child_as_shapeof_v1) {
+                    nodes_for_different_types[index_for_int64].push_back(child);
+                }
+            }
```

Comment:
Please use the same opsets for ShapeOf as mentioned above in the code.
```suggestion
            if (auto shapeof = std::dynamic_pointer_cast<opset3::ShapeOf>(child)) {
                const auto& type_of_output = shapeof ->get_output_type();
                size_t index = (type_of_output == element::i32) ? index_for_int32 : index_for_int64;
                nodes_for_different_types[index].push_back(child);
            } else if (auto shapeof = std::dynamic_pointer_cast<opset1::ShapeOf>(child)) {
                 nodes_for_different_types[index_for_int64].push_back(child);
            }
```

Reply:
BTW, can we use just `child->output_type()` to determine the index without switches?

Reply:
Done.